File: sakila-payment-service/src/main/java/com/example/sakila/module/rental/RentalController.java
package com.example.sakila.module.rental;

import com.example.sakila.exception.NotFoundException;
import com.example.sakila.generated.server.api.RentalsApi;
import com.example.sakila.generated.server.model.RentalDTO;
import com.example.sakila.module.customer.Customer;
import com.example.sakila.module.customer.CustomerService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;

import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

@Controller
public class RentalController implements RentalsApi {

  private final RentalService rentalService;

  private final CustomerService customerService;

  @Autowired
  public RentalController(RentalService rentalService, CustomerService customerService) {
    this.rentalService = rentalService;
    this.customerService = customerService;
  }

  @Override
  public ResponseEntity<RentalDTO> getRentalById(@PathVariable("id") Long id) {
    return ResponseEntity.ok(toDTO(rentalService.getRentalById(id)));
  }

  @Override
  public ResponseEntity<List<RentalDTO>> getRentalsByInventoryId(@PathVariable("id") Long id) {
    return ResponseEntity.ok(
      rentalService.getRentalsByInventoryId(id).stream().map(this::toDTO).collect(Collectors.toList())
    );
  }

  @Override
  public ResponseEntity<List<RentalDTO>> getRentalsByCustomerId(@PathVariable("id") Long id) {
    return ResponseEntity.ok(
        rentalService.getRentalsByCustomerId(id).stream().map(this::toDTO).collect(Collectors.toList())
    );
  }

  @Override
  public ResponseEntity<List<RentalDTO>> getRentalsByStaffId(@PathVariable("id") Long id) {
    return ResponseEntity.ok(
        rentalService.getRentalsByStaffId(id).stream().map(this::toDTO).collect(Collectors.toList())
    );
  }

  @Override
  public ResponseEntity<RentalDTO> createRental(@RequestBody RentalDTO rentalDTO) {
    return ResponseEntity.ok(toDTO(rentalService.createRental(toEntity(rentalDTO))));
  }

  @Override
  public ResponseEntity<RentalDTO> replaceRental(@PathVariable("id") Long id, @RequestBody RentalDTO rentalDTO) {
    return ResponseEntity.ok(toDTO(rentalService.updateRental(id, toEntity(rentalDTO))));
  }

  @Override
  public ResponseEntity<Void> deleteRental(@PathVariable("id") Long id) {
    rentalService.deleteRental(id);
    return ResponseEntity.ok(null);
  }

  private RentalDTO toDTO(Rental rental) {
    RentalDTO rentalDTO = new RentalDTO();
    rentalDTO.setId(rental.getId());
    rentalDTO.setFilmId(rental.getFilmId());
    rentalDTO.setStoreId(rental.getStoreId());
    rentalDTO.setStaffId(rental.getStaffId());
    if (rental.getRentalDate() != null) rentalDTO.setRentalDate(rental.getRentalDate());
    if (rental.getCustomer() != null) rentalDTO.setCustomerId(rental.getCustomer().getId());
    if (rental.getReturnDate() != null) rentalDTO.setReturnDate(rental.getReturnDate());
    if (rental.getLastUpdate() != null) rentalDTO.setLastUpdate(rental.getLastUpdate());
    return rentalDTO;
  }

  private Rental toEntity(RentalDTO rentalDTO) {
    Rental rental = new Rental();
    rental.setId(rentalDTO.getId());
    rental.setStaffId(rentalDTO.getStaffId());
    rental.setFilmId(rentalDTO.getFilmId());
    rental.setStoreId(rentalDTO.getStoreId());

    if (rentalDTO.getCustomerId() != null) {
      Customer customer = customerService.getCustomerById(rentalDTO.getCustomerId());
      if (customer == null) throw new NotFoundException(
          "Custmer for ID " + rentalDTO.getCustomerId() + " does not exist"
      );

      rental.setCustomer(customer);
    }

    if (rentalDTO.getRentalDate() != null) rental.setRentalDate(rentalDTO.getRentalDate());
    if (rentalDTO.getReturnDate() != null) rental.setReturnDate(rentalDTO.getReturnDate());
    if (rentalDTO.getLastUpdate() != null) rental.setLastUpdate(rentalDTO.getLastUpdate());

    return rental;
  }

  private OffsetDateTime toOffsetDateTime(Date date) {
    return OffsetDateTime.ofInstant(date.toInstant(), ZoneId.systemDefault());
  }
}

File: sakila-payment-service/src/main/java/com/example/sakila/module/rental/RentalController.kt
package com.example.sakila.module.rental

import com.example.sakila.exception.BadRequestException
import com.example.sakila.exception.NotFoundException
import com.example.sakila.generated.server.api.RentalsApi
import com.example.sakila.generated.server.model.RentalDTO
import com.example.sakila.module.customer.CustomerService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod

@Controller
class RentalController @Autowired constructor(
    private val rentalService: RentalService,

    private val customerService: CustomerService
) : RentalsApi {

  @RequestMapping(
      value = ["/rentals/{id}"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.GET]
  ) override fun getRentalById(@PathVariable("id") id: Long): ResponseEntity<RentalDTO?> {
    val rental = rentalService.getRentalById(id)?: return ResponseEntity.ok().build()

    return ResponseEntity.ok(toDTO(rental))
  }

  @RequestMapping(
      value = ["/rentals/customer/{id}"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.GET]
  ) override fun getRentalsByCustomerId(@PathVariable("id") id: Long): ResponseEntity<List<RentalDTO>?> {
    val rentals = rentalService.getRentalsByCustomerId(id)?: return ResponseEntity.ok().build()

    return ResponseEntity.ok(rentals.map { toDTO(it) })
  }

  @RequestMapping(
      value = ["/rentals/staff/{id}"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.GET]
  ) override fun getRentalsByStaffId(@PathVariable("id") id: Long): ResponseEntity<List<RentalDTO>?> {
    val rentals = rentalService.getRentalsByStaffId(id)?: return ResponseEntity.ok().build()

    return ResponseEntity.ok(rentals.map { toDTO(it) })
  }

  @RequestMapping(
      value = ["/rentals"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.POST]
  ) override fun createRental(@RequestBody rentalDTO: RentalDTO): ResponseEntity<RentalDTO> {
    var rental = toEntity(rentalDTO)
    rental = rentalService.createRental(rental)

    val result = toDTO(rental)
    return ResponseEntity.ok(result)
  }

  @RequestMapping(
      value = ["/rentals/{id}"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.PUT]
  ) override fun replaceRental(
      @PathVariable("id") id: Long?,
      @RequestBody rentalDTO: RentalDTO
  ): ResponseEntity<RentalDTO> {
    if (id == null) throw BadRequestException("ID should not be null")
    var rental = toEntity(rentalDTO)
    rental = rentalService.updateRental(id, rental)

    val result = toDTO(rental)
    return ResponseEntity.ok(result)
  }

  @RequestMapping(
      value = ["/rentals/{id}"],
      produces = ["application/json"],
      consumes = ["application/json"],
      method = [RequestMethod.DELETE]
  ) override fun deleteRental(@PathVariable("id") id: Long?): ResponseEntity<Void> {
    if (id == null) throw BadRequestException("ID should not be null")
    rentalService.deleteRental(id)
    return ResponseEntity.ok().build()
  }

  private fun toDTO(rental: Rental): RentalDTO {
    val rentalDTO = RentalDTO()

    rentalDTO.id = rental.id
    rentalDTO.filmId = rental.filmId
    rentalDTO.storeId = rental.storeId
    rentalDTO.staffId = rental.staffId
    rentalDTO.rentalDate = rental.rentalDate
    rentalDTO.returnDate = rental.returnDate
    rentalDTO.customerId = rental.customer?.id
    rentalDTO.lastUpdate = rental.lastUpdate

    return rentalDTO
  }

  private fun toEntity(rentalDTO: RentalDTO): Rental {
    val rental = Rental()

    rental.id = rentalDTO.id
    rental.staffId = rentalDTO.staffId
    rental.filmId = rentalDTO.filmId
    rental.storeId = rentalDTO.storeId
    rental.returnDate = rentalDTO.returnDate
    rental.rentalDate = rentalDTO.rentalDate
    rental.lastUpdate = rentalDTO.lastUpdate

    if (rentalDTO.customerId != null) {
      val customerId = rentalDTO.customerId
      val customer = customerService.getCustomerById(customerId)?:
        throw NotFoundException("Customer for ID $customerId does not exist")

      rental.customer = customer
    }

    return rental
  }

}
